src/peft/tuners/ia3.py

```diff
@@ -532,21 +532,21 @@ def __init__(
             self.is_feedforward = is_feedforward
 
         def forward(self, x: torch.Tensor):
-            if self.disable_adapters or self.active_adapter not in self.ia3_l.keys():
+            if self.disable_adapters or (self.active_adapter not in self.ia3_l.keys()):
                 return super().forward(x)
+
+            requires_conversion = (not torch.is_autocast_enabled()) and (x.dtype != torch.float32)
+            if requires_conversion:
+                x = x.float()
+
+            ia3_scaling = self.ia3_l[self.active_adapter].flatten()
+            if self.is_feedforward:
+                result = super().forward(x * ia3_scaling)
             else:
-                if not torch.is_autocast_enabled():
-                    if x.dtype != torch.float32:
-                        x = x.float()
-                    if self.is_feedforward:
-                        result = super().forward(x * self.ia3_l[self.active_adapter].flatten())
-                    else:
-                        result = super().forward(x)
-                        expected_dtype = result.dtype
-                        result = (result * self.ia3_l[self.active_adapter].flatten()).to(expected_dtype)
-                else:
-                    if self.is_feedforward:
-                        result = super().forward(x * self.ia3_l[self.active_adapter].flatten())
-                    else:
-                        result = result * self.ia3_l[self.active_adapter].flatten()
+                result = super().forward(x)
+                expected_dtype = result.dtype
+                result = result * ia3_scaling
+
+            if requires_conversion:
+                result = result.to(expected_dtype)
```

Comment:
```suggestion
                if requires_conversion:
                    result = result.to(expected_dtype)
```

Oh, I see a bug in my proposed code. `expected_dtype` might not exist, so this either needs to be indented (as suggested), leading to 2 levels of indentation ;-( or we would have to do something like:

```python
# near the top of the function
expected_dtype = None

...

if requires_conversion and expected_dtype is not None:
    result = result.to(expected_dtype)
```

However, there is still this issue I mentioned earlier:

> I'm not sure if it's correct that the dtype-conversion is not applied when `self.is_feedforward`, but that's already the case with the existing code.

@SumanthRH @pacman100